lib/DxilValidation/DxilValidation.cpp

```diff
@@ -970,6 +970,169 @@ static void ValidateImmOperandForMathDxilOp(CallInst *CI, DXIL::OpCode opcode,
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrMatVecOpIsUnsignedFlagsAreConst);
+  }
+
+  if (!llvm::isa<llvm::Constant>(InputInterpretation) ||
+      !llvm::isa<llvm::Constant>(MatrixInterpretation)) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgInterpretationParamAreConst);
+  }
+
+  // Check if InputInterpretation and MatrixInterpretation are valid
+  ConstantInt *Ii = cast<ConstantInt>(InputInterpretation);
+  auto IiValue = Ii->getLimitedValue();
+  if (!CheckFromRegisterInterpretations(IiValue)) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgInvalidRegisterInterpValue);
+  }
+
+  ConstantInt *Mi = cast<ConstantInt>(MatrixInterpretation);
+  auto MiValue = Mi->getLimitedValue();
+  if (!CheckInMemoryInterpretations(MiValue)) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgInvalidMemoryInterpValue);
+  }
+
+  if (!llvm::isa<llvm::Constant>(MatrixM) ||
+      !llvm::isa<llvm::Constant>(MatrixK) ||
+      !llvm::isa<llvm::Constant>(MatrixLayout) ||
+      !llvm::isa<llvm::Constant>(MatrixTranspose)) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgMatrixShapeParamsAreConst);
+  }
+
+  ConstantInt *Ml = cast<ConstantInt>(MatrixLayout);
+  auto MlValue = Ml->getLimitedValue();
+  if (!CheckMatrixLayout(MlValue)) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgInvalidMatrixLayoutValue);
+  }
+
+  switch (opcode) {
+  case DXIL::OpCode::MatVecMul: {
+    llvm::Value *OutputIsUnsigned =
+        CI->getOperand(DXIL::OperandIndex::kMatVecMulIsOutputUnsignedIdx);
+    if (!llvm::isa<llvm::Constant>(OutputIsUnsigned)) {
+      ValCtx.EmitInstrError(
+          CI, ValidationRule::InstrMatVecOpIsUnsignedFlagsAreConst);
+    }
+
+  } break;
+  case DXIL::OpCode::MatVecMulAdd: {
+    llvm::Value *OutputIsUnsigned =
+        CI->getOperand(DXIL::OperandIndex::kMatVecMulAddIsOutputUnsignedIdx);
+    if (!llvm::isa<llvm::Constant>(OutputIsUnsigned)) {
+      ValCtx.EmitInstrError(
+          CI, ValidationRule::InstrMatVecOpIsUnsignedFlagsAreConst);
+    }
+    llvm::Value *BiasInterpretation =
+        CI->getOperand(DXIL::OperandIndex::kMatVecMulAddBiasInterpretation);
+    if (!llvm::isa<llvm::Constant>(BiasInterpretation)) {
+      ValCtx.EmitInstrError(
+          CI, ValidationRule::InstrLinalgInterpretationParamAreConst);
+    }
+    ConstantInt *Bi = cast<ConstantInt>(BiasInterpretation);
+    auto BiValue = Bi->getLimitedValue();
+    if (!CheckInMemoryInterpretations(BiValue)) {
+      ValCtx.EmitInstrError(
+          CI, ValidationRule::InstrLinalgInvalidMemoryInterpValue);
+    }
+  } break;
+  default:
+    break;
+  }
+}
+
+static void ValidateImmOperandsForOuterProdAcc(CallInst *CI,
+                                               DXIL::OpCode opcode,
+                                               ValidationContext &ValCtx) {
```

Comment:
I believe `opcode` is unused, right?
```suggestion
static void ValidateImmOperandsForOuterProdAcc(CallInst *CI,
                                               ValidationContext &ValCtx) {
```